### PR TITLE
Add autocorrection to MarkRule

### PR DESCRIPTION
**Pull request by kohtenko**
This pull request about adding autocorrect functionality to newly added MarkRule, in order to make updating code even simpler.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/MarkRule.swift [hunk 1]
```diff
@@ -39,20 +39,112 @@ public struct MarkRule: ConfigurationProviderRule {
         ]
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/MarkRule.swift]: please add `corrections` to this `RuleDescription` to cover correctable test scenarios.

File: Source/SwiftLintFramework/Rules/MarkRule.swift [hunk 2]
```diff
@@ -39,20 +39,112 @@ public struct MarkRule: ConfigurationProviderRule {
...
+
+    private var twoOrMoreSpacesAfterHyphenPattern: String {
+        return "(\(mark) -\(twoOrMoreSpace))"
+    }
+
+    private var nonSpaceAfterHyphenPattern: String {
+        return "(\(mark) -\(nonSpace))"
+    }
+
+    private var pattern: String {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/MarkRule.swift]: Suggested formatting:

```swift
private var pattern: String {
    return [
        spaceStartPattern,
        endNonSpacePattern,
        endTwoOrMoreSpacePattern,
        twoOrMoreSpacesAfterHyphenPattern,
        nonSpaceAfterHyphenPattern
    ].joinWithSeparator("|")
}
```

File: Source/SwiftLintFramework/Rules/MarkRule.swift [hunk 3]
```diff
@@ -39,20 +39,112 @@ public struct MarkRule: ConfigurationProviderRule {
...
+            let location = Location(file: file, characterOffset: range.location)
+            nsstring = nsstring.stringByReplacingCharactersInRange(range, withString: replaceString)
+            corrections.append(Correction(ruleDescription: description, location: location))
+        }
+        file.write(nsstring as String)
+        file.invalidateCache()
+        return corrections
+    }
+
+    private func violationRangesInFile(file: File, withPattern pattern: String) -> [NSRange] {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/MarkRule.swift]: Suggested formatting:

```swift
private func violationRangesInFile(file: File, withPattern pattern: String) -> [NSRange] {
    let nsstring = file.contents as NSString
    return file.rangesAndTokensMatching(pattern).filter { range, syntaxTokens in
        let syntaxKinds = syntaxTokens.flatMap { SyntaxKind(rawValue: $0.type) }
        return syntaxKinds.startsWith([.Comment])
    }.flatMap { range, syntaxTokens in
        let identifierRange = nsstring
            .byteRangeToNSRange(start: syntaxTokens[0].offset, length: 0)
        return identifierRange.map { NSUnionRange($0, range) }
    }
}
```

File: Source/SwiftLintFramework/Rules/MarkRule.swift [hunk 4]
```diff
@@ -39,20 +39,112 @@ public struct MarkRule: ConfigurationProviderRule {
...
+        let violations = violationRangesInFile(file, withPattern: pattern)
+        let matches = file.ruleEnabledViolatingRanges(violations, forRule: self)
+        if matches.isEmpty { return [] }
+
+        var nsstring = file.contents as NSString
+        let description = self.dynamicType.description
+        var corrections = [Correction]()
+        for var range in matches.reverse() {
+            if keepLastChar {
+                range.length = range.length - 1
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/MarkRule.swift]: `range.length -= 1`?

File: Source/SwiftLintFramework/Rules/MarkRule.swift [hunk 5]
```diff
@@ -39,20 +39,112 @@ public struct MarkRule: ConfigurationProviderRule {
...
+
+        result.appendContentsOf(correctFile(file,
+            pattern: nonSpaceAfterHyphenPattern,
+            replaceString: "// MARK: - ",
+            keepLastChar: true))
+
+        return result
+    }
+
+    public func correctFile(file: File,
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/MarkRule.swift]: `private`?

**Comment by kohtenko** [comment]
Good points, all fixed

**Pull request merged**